Refactor message edit code and replace the `Never` option in dropdown with a checkbox.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
Description of individual commits in order-
- Renames the message editing section in settings.
- Call `do_set_realm_property` only if setting is actually changed.
- Make `zero` invalid value for `message_content_edit_limit_seconds`.
- Add checkbox for `Allow message editing` setting, and remove `Never` option from time limit dropdown.
- Remove duplicate code in tests.
- Use same list for edit and delete limit dropdown values.
- Disable the `Save changes` button in case of invalid value.
 
Fixes: <!-- Issue link, or clear description.--> parts of #21739.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![message-edit-new](https://user-images.githubusercontent.com/35494118/163856129-6c0ee001-1e85-43e7-a017-176c6d91a4fd.gif)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] Self-reviewed the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on zerver/views/realm.py:
```diff
@@ -221,18 +222,25 @@ def update_realm(
-            message_content_edit_limit_seconds is not None
-            and realm.message_content_edit_limit_seconds != message_content_edit_limit_seconds
+            message_content_edit_limit_seconds_raw is not None
+            and message_content_edit_limit_seconds != realm.message_content_edit_limit_seconds
```
I think this is not clean and we can improve this. We will make the topic editing permissions independent, so one condition will be removed then.

Also, I think we do not need to always update `allow_message_editing` and `message_content_edit_limit_seconds` together, since these are no longer controlled by single element in the UI. This settings are of course dependent, since the limit only matters if `allow_message_editing` is true, but I think we can remove this and handle it like other settings.

Also while sending events, we send all the settings in the object, even when those are not updated. It may makes sense when changing `allow_message_editing` from `False` to `True` but not in other cases. But I am not much sure about this.

sahil839 replied:
@timabbott what are your thoughts on this? Would be great to have feedback on this since working on further parts will depend on how we proceed with this.

sahil839 commented on static/js/settings_org.js:
```diff
@@ -860,28 +865,21 @@ export function register_save_discard_widget_handlers(
                         break;
                     }
                     default: {
-                        data.allow_message_editing = true;
```
I think now we can also remove this and send only changed values the API and not all settings in subsection, as we no longer control multiple settings using single UI element. Though some settings are disabled according to others but that is the UI part, and we need not send all the settings to the API.

timabbott replied:
That sounds correct.

sahil839 commented on static/js/settings_org.js:
```diff
@@ -1088,6 +1091,15 @@ export function build_page() {
+                $("#id_realm_message_content_edit_limit_minutes")
+                    .closest(".org-subsection-parent")
+                    .find(".subsection-changes-save button")
+                    .prop("disabled", true);
```
This is added because we currently show `0` in the input, when setting is cahnged from `anytime` to `custom`. One option can be we can just leave the box empty in that case.

sahil839 commented:
Some points regarding the last commit to disable the save button -
- The event is fired only after focus is out of the input box and not when input is changed to 0.
- Do we want to change the visual look of the `Save changes` button in disabled state?
- The button is disabled initially on changing from `anytime` to `custom` since the value in the input is 0. We can avoid this by changing the value in the input to be something else as mentioned [here](https://github.com/zulip/zulip/pull/21837#discussion_r852322231).
- We only check for 0 value and for other invalid values like letters, we still show the error returned by server as we do for other input box. We can fix this by changing the type of input element to `number` but this results in the arrows for increasing/decreasing the input as shown in the example [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number). We can use some browser specific CSS to hide these arrows but I am not sure we prefer to do that. (Though we use this one of the portico or billing pages I guess).

sahil839 commented:
Note that we still need to update the documentation to mention the new checkbox setting.

alya commented:
Oh, heh, I just saw that you commented on the above as well!

alya commented:
We can try it with the arrows -- it might be OK.

alya commented:
> Do we want to change the visual look of the Save changes button in disabled state?

Yeah, I think greying it out as you've done is good. We do that for disabled buttons in our modals.

sahil839 commented:
@alya Ok, added a separate commit for using `type=number` input. A user is not allowed to type alphabets and most of the other characters. A user can type `-10`, or `1.99` for example, but the input border will turn into red as in custom profile fields. Note that a user can still click on save changes in such case. Do we also want to disable the `Save` button in these cases? I think we can do it.

Also, have kept the minimum value of input to be 0 for now, since otherwise changing the dropdown from `Anytime` to `Custom`  results in a red input border with value as 0 and seems confusing for the user to see red border even when user has not changed anything inside the input. Do we instead want to leave the input empty in that case? For other options like when changing from `1 hour` option to `Custom` the input has "60" in it by default

![msg-edit-custom](https://user-images.githubusercontent.com/35494118/164282502-753930b5-67bd-4de5-88a6-959eec954a4e.gif)

sahil839 commented:
> @alya Ok, added a separate commit for using `type=number` input. A user is not allowed to type alphabets and most of the other characters. A user can type `-10`, or `1.99` for example, but the input border will turn into red as in custom profile fields. Note that a user can still click on save changes in such case. Do we also want to disable the `Save` button in these cases? I think we can do it.
> 
> Also, have kept the minimum value of input to be 0 for now, since otherwise changing the dropdown from `Anytime` to `Custom` results in a red input border with value as 0 and seems confusing for the user to see red border even when user has not changed anything inside the input. Do we instead want to leave the input empty in that case? For other options like when changing from `1 hour` option to `Custom` the input has "60" in it by default
> 
> ![msg-edit-custom](https://user-images.githubusercontent.com/35494118/164282502-753930b5-67bd-4de5-88a6-959eec954a4e.gif)

@alya also added the commit to disable delete limit dropdown without save. Would ping Tim for the code review and the code cleanup questions once you feel that the UI changes are fine.

alya commented:
> @alya Ok, added a separate commit for using `type=number` input. A user is not allowed to type alphabets and most of the other characters. A user can type `-10`, or `1.99` for example, but the input border will turn into red as in custom profile fields. Note that a user can still click on save changes in such case. Do we also want to disable the `Save` button in these cases? I think we can do it.

Yeah, I think we should disable "Save" any time that clicking it would produce an error.

alya commented:
I was also able to get a couple of error messages by typing large numbers into the field.

Moderately large number:
![Screen Shot 2022-06-03 at 12 05 08 AM](https://user-images.githubusercontent.com/2090066/171805246-e1a009dd-cf28-4503-954a-ca9f790b3af7.png)

Even larger number:
![Screen Shot 2022-06-03 at 12 04 33 AM](https://user-images.githubusercontent.com/2090066/171805415-68001fad-7cdb-442c-83d7-5be9009cfe11.png)

I think it would be good to highlight in red / disable "Save" in these cases as well.

alya commented:
> Also, have kept the minimum value of input to be 0 for now, since otherwise changing the dropdown from `Anytime` to `Custom` results in a red input border with value as 0 and seems confusing for the user to see red border even when user has not changed anything inside the input. Do we instead want to leave the input empty in that case? For other options like when changing from `1 hour` option to `Custom` the input has "60" in it by default

Mm, yeah, let's leave the input empty and highlight 0 in red when switching from `Anytime`. I think that's a fine solution.

alya commented:
I think we need to make the box for entering the custom time larger. It should probably fit the number of minutes in a week, as the user will see it when switching from `1 week` to `Custom`.

Could we also make it less cramped on the right? At least for me (Chrome on a Mac), it looks like this:

![Screen Shot 2022-06-03 at 12 03 19 AM](https://user-images.githubusercontent.com/2090066/171806631-1a58d6fb-5a68-429f-a8b8-ca4429d5cef6.png)
![Screen Shot 2022-06-03 at 12 03 32 AM](https://user-images.githubusercontent.com/2090066/171806639-cf307e35-5af3-4679-937d-8b95e281916a.png)

alya commented:
Sorry for the delayed review, and please let me know if I missed any open questions!

sahil839 commented:
Added commits for -
- Inreasing the width of input.
- "0" value will make the input box red and changing from "Anytime" to "Custom" will make the input empty.

Regarding other changes -
- Do we want to keep any maximum limit for these time limits? Currently any value above the range which is accepted by django field returns error. The django `PostiveIntegerField` can have maximum value `2147483647` (which is in seconds), so the maximum value of the custom input currently is the floor of `2147483647/60`, i.e. `35791394`, and anything above that will return error. So, we can add a maximum value to input once we decide what we want the maximum value to be.
- Regarding the arrows in the input box, we can add some margin between the text and arrows, but the CSS for this would be written using classes like `input::-webkit-outer-spin-button` as mentioned in `https://www.w3schools.com/howto/howto_css_hide_arrow_number.asp`. So, if we want to use these classes, we can instead hide the arrows too as mentiond in the shared link.

alya commented:
I don't have a strong perspective on the maximum value. 35791394 might be fine, or maybe something like 525600 (= 1 year). @timabbott thoughts?

alya commented:
I'm OK with the arrows looking a bit janky, since it's not a very prominent part of the UI, if that makes it significantly simple from an eng perspective.

alya commented:
A couple more observations:

1. If you change an invalid input (e.g. empty box) to a valid input, the "Save" button is still disabled until you click outside the input field, which seems confusing. I'd expect it to change as you type.

<img width="546" alt="Screen Shot 2022-06-03 at 1 51 34 PM" src="https://user-images.githubusercontent.com/2090066/171950962-aad334cb-8f66-41d3-89d5-683181520d93.png">

2. I can still get a non-user-facing error message by deleting all the text in the input box. Perhaps getting the button state to update properly as you type would fix this issue as well?

<img width="546" alt="Screen Shot 2022-06-03 at 1 51 53 PM" src="https://user-images.githubusercontent.com/2090066/171951224-ff577f7b-3654-4e6c-ae63-c3912d0ac78a.png">

alya commented:
This is surprisingly tricky!

sahil839 commented:
I updated the PR to disable/enable the button while typing in the input box and also it is disabled if the input box is empty. Still waiting for Tim's feedback on the maximum value question to make the changes, I think 1 year could also be fine, in that case we need to implement it at the API level also.

alya commented:
Great, thanks! I think getting feedback from @timabbott is the next step at this point.

sahil839 commented:
@timabbott can you review this when you are available? Its review and feedback on some of the above question will help in doing next part of the issue.

sahil839 commented:
@timabbott this PR is ready to review.

timabbott commented:
So, I'm uncertain about the direction of the last ~5 commits. There's still bugs, like if you enter 30 digits, you get this error:

![image](https://user-images.githubusercontent.com/2746074/184254445-bce5ccd2-0338-4573-ae88-bf318b441ef5.png)

(Apparently, because JS turns it into `1.11231312313213131e+29`). 

And the HTML integer up/down buttons are confusing in this context; they're a UI element designed for "How many apples do you want" type elements where incrementing/decrementing by 1 makes sense.

![image](https://user-images.githubusercontent.com/2746074/184254587-a8fda282-6454-43bd-a7a7-260f450ab7ee.png)

At the same time, we've added 80 lines of custom JS code just for these inputs to try to work around some of the annoying issues for them discussed above, and I'm not just not sure it's worth the investment/maintenance/effort over just having these be string inputs where you're clearly supposed to enter a number of minutes, and we try to make sure the error messages are OK.

I'd really like to merge all the other work in this PR; @sahil839 is it easy to split the commits related to making it an `integer` input to a separate pull request and have a state that's clearly better than `main`? I'm not sure which of those last 5 commits are part of that idea or other things -- the commit messages for those aren't super clear about what the actual bug is that they're addressing.

sahil839 commented:
I have removed the commit for allowing only `integer` in the input. But we still have commit for disabling the save button if input is invalid, i.e. 0, -ve or empty. Note that if the input is `111aa`, then `Number.parseInt` still converts it to `111` and it is treated as valid.

timabbott commented on static/js/settings_org.js:
```diff
@@ -1197,6 +1197,22 @@ export function build_page() {
+            "id_realm_message_content_delete_limit_minutes",
+            true,
+        );
+    });
```
Could this share code with `set_delete_own_message_policy_dropdown`? I noticed that's the pattern for the similar `#id_realm_allow_message_editing`.

timabbott commented:
I rebased this to address the merge conflicts/renumbering and marked this to merge once CI passes.

I figure if we can deduplicate the bit of code I mentioned above, that can be a small follow-up.

Huge thanks for doing this transition @sahil839!
